chainer/optimizers/adam.py

```diff
@@ -144,96 +199,20 @@ def init_state(self, param):
             self.state['v'] = intel64.ideep.array(
                 self.state['v'], itype=intel64.ideep.wgt_array)
 
-    def _check_eps(self, interm_dtype):
-        # Checks that the eps does not underflow.
-        hp = self.hyperparam
-        eps = interm_dtype(hp.eps)
-        if hp.eps != 0 and eps == 0:
-            raise ValueError(
-                'eps of Adam optimizer is too small for {} ({})'.format(
-                    interm_dtype.name, hp.eps))
-        # Note that the converted `eps` (numpy scalar) is discarded here and
-        # the original `hp.eps` is used in calculation, because Python
-        # scalars are faster in cupy elementwise kernels.
-
-    def update_core_cpu(self, param):
+    def _update_core_cpu_or_gpu(self, param):
```

Comment:
How about overriding `update_core`?

Reply:
I agree. Could you fix this?

Reply:
I want to call `update_core_chainerx` when `chainerx.ndarray` are given as parameters. Do you have any ideas?

Reply:
You can call `update_core_chainerx` from `update_core` by yourself, just as in `UpdateRule.update_core`.

Reply:
It will not work well if we just rewrite `update_core` because `update_core_chainerx` calls `update_core_cpu` or `update_core_gpu` directly.

Reply:
Rebased and fixed to just rewrite `update_core`.